pkg/event-consumers/kafka/kafka-consumer.go

```diff
@@ -0,0 +1,155 @@
+		return fmt.Errorf("Unable to find the service for function %s", funcName)
+	}
+	logrus.Infof("Sending message %s to function %s", msg, funcName)
+	funcPort := strconv.Itoa(int(svc.Spec.Ports[0].Port))
+
+	req, err := http.NewRequest("POST", fmt.Sprintf("http://%s.%s.svc.cluster.local:%s", funcName, ns, funcPort), strings.NewReader(msg))
+	if err != nil {
+		return fmt.Errorf("Failed to create a request %v", req)
+	}
+	timestamp := time.Now().UTC()
+	req.Header.Add("Content-Type", "application/x-www-form-urlencoded")
+	req.Header.Add("event-id", fmt.Sprintf("kafka-consumer-%s-%s-%s", funcName, ns, timestamp.Format(time.RFC3339Nano)))
+	req.Header.Add("event-type", "application/x-www-form-urlencoded")
+	req.Header.Add("event-time", timestamp.String())
+	req.Header.Add("event-namespace", "kafkatriggers.kubeless.io")
+	client := &http.Client{}
+	_, err = client.Do(req)
+	if err != nil {
+		//detect the request timeout case
+		if strings.Contains(err.Error(), "status code 408") {
+			return errors.New("Request timeout exceeded")
+		}
+		return err
+	}
+
+	logrus.Infof("Message has sent to function %s successfully", funcName)
+	return nil
+}
+
+// CreateKafkaConsumer creates a goroutine that subscribes to Kafka topic
+func CreateKafkaConsumer(triggerObjName, funcName, ns, topic string) error {
+	consumerID := generateUniqueConsumerGroupID(triggerObjName, funcName, ns, topic)
+	if !consumerM[consumerID] {
+		logrus.Infof("Creating Kafka consumer for the function %s associated with for trigger %s", funcName, triggerObjName)
+		stopM[consumerID] = make(chan struct{})
+		stoppedM[consumerID] = make(chan struct{})
+		go createConsumerProcess(brokers, topic, funcName, ns, consumerID, stopM[consumerID], stoppedM[consumerID])
+		consumerM[consumerID] = true
+		logrus.Infof("Created Kafka consumer for the function %s associated with for trigger %s", funcName, triggerObjName)
+	} else {
+		logrus.Infof("Consumer for function %s associated with trigger %s already exists, so just returning", funcName, triggerObjName)
+	}
+	return nil
+}
+
+// DeleteKafkaConsumer deletes goroutine created by CreateKafkaConsumer
+func DeleteKafkaConsumer(triggerObjName, funcName, ns, topic string) error {
+	consumerID := generateUniqueConsumerGroupID(triggerObjName, funcName, ns, topic)
+	if consumerM[consumerID] {
+		logrus.Infof("Stopping consumer for the function %s associated with for trigger %s", funcName, triggerObjName)
+		// delete consumer process
+		close(stopM[consumerID])
+		<-stoppedM[consumerID]
+		consumerM[consumerID] = false
+		logrus.Infof("Stopped consumer for the function %s associated with for trigger %s", funcName, triggerObjName)
+	} else {
+		logrus.Infof("Consumer for function %s associated with trigger does n't exists. Good enough to skip the stop", funcName, triggerObjName)
+	}
+	return nil
+}
+
+func generateUniqueConsumerGroupID(triggerObjName, funcName, ns, topic string) string {
+	return ns + "_" + triggerObjName + "_" + funcName + "_" + topic
```

Comment:
> less descriptive, but is trigger.UID sufficient for a unique consumer group?

yes, it would be the same

> In particular, if I delete and recreate the Trigger (instead of update), do I want to receive events for the old Trigger?

I am not sure I am understanding this but if you delete the trigger you won't receive any more messages.